// json-decode/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    array: function() {
        return _jsondecode.array;
    },
    bigint: function() {
        return _jsondecode.bigint;
    },
    bool: function() {
        return _jsondecode.bool;
    },
    Decoder: function() {
        return _jsondecode.Decoder;
    },
    DecoderError: function() {
        return _jsondecode.DecoderError;
    },
    int: function() {
        return _jsondecode.int;
    },
    field: function() {
        return _jsondecode.field;
    },
    number: function() {
        return _jsondecode.number;
    },
    float: function() {
        return _jsondecode.float;
    },
    nullable: function() {
        return _jsondecode.nullable;
    },
    optional: function() {
        return _jsondecode.optional;
    },
    string: function() {
        return _jsondecode.string;
    }
});
var _jsondecode = require("./lib/json-decode");
